### PR TITLE
Fix #87 by avoiding double error log

### LINKED ISSUE #87: Action.onError is called twice if it throws an exception
UniFlow 1.0.10 calls `Action.onError` twice for error functions that throw an exception. The default `onError` function for data flows throws an exception, defined here: https://github.com/uniflow-kt/uniflow-kt/blob/1.0.10/uniflow-core/src/main/kotlin/io/uniflow/core/flow/DataFlow.kt#L42-L45. So, the default behaviour is that the error function is run twice.

Likely, this is caused by `actionOn<T : UIState> {}` being reduced by https://github.com/uniflow-kt/uniflow-kt/blob/1.0.10/uniflow-core/src/main/kotlin/io/uniflow/core/flow/ActionReducer.kt#L50-L62. This code calls `action.onError` twice, because:

1. The state isn't the target state: 👍.
2. The `BadOrWrongStateException` is thrown by the first `onError` call, which is caught by `catch (e: Exception)` which runs `onError` again: 👎.

The MWE below simply tests that the default `onError` function is called twice, because the default function logs a message. The test simply counts the logged error messages. There should be exactly one, but there are two. This is by no means a good test for this issue, but it simply demonstrates the issue.

MWE (deps: JUnit 4.13.2, UniFlow 1.0.10):

```kotlin
import io.uniflow.android.AndroidDataFlow
import io.uniflow.core.flow.actionOn
import io.uniflow.core.flow.data.UIEvent
import io.uniflow.core.flow.data.UIState
import io.uniflow.core.logger.Logger
import io.uniflow.core.logger.UniFlowLogger
import io.uniflow.test.rule.UniflowTestDispatchersRule
import kotlin.test.assertEquals
import kotlinx.coroutines.test.TestCoroutineDispatcher
import org.junit.Rule
import org.junit.Test

class SingleLogTest {
    companion object {
        private val testCoroutineDispatcher = TestCoroutineDispatcher()
    }

    @get:Rule
    val uniflowTestDispatchersRule = UniflowTestDispatchersRule(testCoroutineDispatcher)

    @Test
    fun `bad or wrong state is logged once`() {
        val logger = object : Logger {
            val messages = mutableListOf<String>()

            override fun debug(message: String) {
            }

            override fun log(message: String) {
            }

            override fun logError(errorMessage: String, error: Exception?) {
                messages += errorMessage
                println(errorMessage)
            }

            override fun logEvent(event: UIEvent) {
            }

            override fun logState(state: UIState) {
            }
        }

        UniFlowLogger.init(logger)

        val androidDataFlow = object : AndroidDataFlow(defaultState = UIState.Empty) {
            fun foo() = actionOn<UIState.Loading> {}
        }

        androidDataFlow.foo() // Should log an uncaught error about a BadOrWrongStateException

        assertEquals(1, logger.messages.size)
    }
}
```

Fails with

```
expected:<1> but was:<2>
Expected :1
Actual   :2
```

### PR DESCRIPTION
**Pull request by arnaudgiuliani**
This PR is fixing double error logging when having `BadOrWrongStateException`

### CODE REVIEW

File: uniflow-android-test/src/test/java/io/uniflow/android/test/sync/OtherTests.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        UniFlowLogger.init(logger)
+
+        val androidDataFlow = object : AndroidDataFlow(defaultState = UIState.Empty) {
+            fun foo() = actionOn<UIState.Loading> {}
+        }
+
+        androidDataFlow.foo() // Should log an uncaught error about a BadOrWrongStateException
+
+        assertEquals(1, logger.messages.size)
```
Comment (erikhuizinga) [inline uniflow-android-test/src/test/java/io/uniflow/android/test/sync/OtherTests.kt:56]:
```suggestion
        assertEquals(1, logger.messages.size)
        
        UniFlowLogger.default()
```

**Review by erikhuizinga — APPROVED** [review]
You probably want to reset to the default uniflow logger after this test.
LGTM otherwise! 👍

File: uniflow-android-test/src/test/java/io/uniflow/android/test/sync/OtherTests.kt (hunk 1 above)
Comment (arnaudgiuliani) [reply uniflow-android-test/src/test/java/io/uniflow/android/test/sync/OtherTests.kt:56]: yep

**Comment by arnaudgiuliani** [comment]
Great, thanks

**Pull request merged**